core/src/SYCL/Kokkos_SYCL_Parallel_Scan.hpp

```diff
@@ -0,0 +1,324 @@
+                               functor, &global_mem[global_id],
+                               &group_results[item.get_group_linear_id()]);
+                       });
+    });
+    q.wait();
+  }
+
+  template <typename Functor>
+  void sycl_direct_launch(const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    cl::sycl::queue& q = *instance.m_queue;
+
+    const std::size_t len = m_policy.end() - m_policy.begin();
+
+    // Initialize global memory
+    q.submit([&](sycl::handler& cgh) {
+      auto global_mem = m_scratch_space;
+      auto policy     = m_policy;
+      cgh.parallel_for(sycl::range<1>(len), [=](sycl::item<1> item) {
+        const typename Policy::index_type id =
+            static_cast<typename Policy::index_type>(item.get_id()) +
+            policy.begin();
+        value_type update{};
+        ValueInit::init(functor, &update);
+        if constexpr (std::is_same<WorkTag, void>::value)
+          functor(id, update, false);
+        else
+          functor(WorkTag(), id, update, false);
+        ValueOps::copy(functor, &global_mem[id], &update);
+      });
+    });
+    q.wait();
+
+    // Perform the actual exlcusive scan
+    scan_internal(q, functor, m_scratch_space, len);
+
+    // Write results to global memory
+    q.submit([&](sycl::handler& cgh) {
+      auto global_mem = m_scratch_space;
+      cgh.parallel_for(sycl::range<1>(len), [=](sycl::item<1> item) {
+        auto global_id = item.get_id();
+
+        value_type update = global_mem[global_id];
+        if constexpr (std::is_same<WorkTag, void>::value)
+          functor(global_id, update, true);
+        else
+          functor(WorkTag(), global_id, update, true);
+        ValueOps::copy(functor, &global_mem[global_id], &update);
+      });
+    });
+    q.wait();
+  }
+
+  template <typename Functor>
+  void sycl_indirect_launch(const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
+        *instance.m_indirectKernel;
+
+    // Allocate USM shared memory for the functor
+    kernelMem.resize(std::max(kernelMem.size(), sizeof(functor)));
+
+    // Placement new a copy of functor into USM shared memory
+    //
+    // Store it in a unique_ptr to call its destructor on scope exit
+    std::unique_ptr<Functor, Kokkos::Impl::destruct_delete> kernelFunctorPtr(
+        new (kernelMem.data()) Functor(functor));
+
+    auto kernelFunctor = std::reference_wrapper(*kernelFunctorPtr);
+    sycl_direct_launch(kernelFunctor);
+  }
+
+ public:
+  template <typename PostFunctor>
+  void impl_execute(PostFunctor&& post_functor) {
```

Comment:
Why do we use a forwarding reference here?

Reply:
When not forwarding, `const &` is actually better, because there is a corner case if the functor has both `operator()() const` and `operator()()`, how you pass it determines which one gets called, which is really subtle.

Reply:
Addressed in [`8389c54` (#3577)](https://github.com/kokkos/kokkos/pull/3577/commits/8389c54be0d766d7c941c6cf8fa1b2ffe49b355b).